### PR TITLE
[core] More vertex buffer validation

### LINKED ISSUE #6799: Validate `VertexBufferLayout::array_stride`
### Context

The WebGPU [spec](https://gpuweb.github.io/gpuweb/#dictdef-gpuvertexbufferlayout) constraints what the array stride of a `VertexBufferLayout` might be. In particular, it forbids it to be smaller or equal than the `max(offset + size)` over all attributes:

image

One might be tempted to use a stride smaller than the attributes to implement some kind of "sliding window" over a vertex buffer (I certainly was). This is however non-compliant as per above, and may variously lead to at least:
- recent chrome versions to reject your shader
- wgpu 23.0.0+ to emit incorrect metal shader (including out-of-bound array access)
- possibly other phenomenons (I have only tested mac, WebGPU, and WebGL, the latter seemingly being unaffected)

### Issue

Currently, wgpu doesn't warn/error on non-compliant stride values. Since 23.0.0,  it also silently generate erroneous metal shader when provided with non-compliant stride.

### Desired change

wgpu should error (or warn loudly) with a reference to the spec whenever a non-compliant stride is detected.

### Reference

- PR which lead to the discovery of the above with my debugging notes, in particular the offending metal shader: https://github.com/abey79/vsvg/pull/160
- PR in which is switch from a sliding window to a compliant, multiple-buffer-binding-with-offset ("zip"):  https://github.com/abey79/vsvg/pull/159 (This may inspire a work-around to anyone facing this issue.)

### PR DESCRIPTION
**Pull request by sagudev**
**Connections**
Fix #6799

**Testing**
[CTS run in servo](https://github.com/sagudev/servo/actions/runs/12452023550/job/34761648385):
```
OK /_webgpu/webgpu/cts.https.html?q=webgpu:api,validation,capability_checks,limits,maxVertexAttributes:createRenderPipeline,at_over:*

    PASS [expected FAIL] subtest: :limitTest="atDefault";testValueName="overLimit";async=false
    PASS [expected FAIL] subtest: :limitTest="atDefault";testValueName="overLimit";async=true
    PASS [expected FAIL] subtest: :limitTest="underDefault";testValueName="overLimit";async=false
    PASS [expected FAIL] subtest: :limitTest="underDefault";testValueName="overLimit";async=true
    PASS [expected FAIL] subtest: :limitTest="betweenDefaultAndMaximum";testValueName="overLimit";async=false
    PASS [expected FAIL] subtest: :limitTest="betweenDefaultAndMaximum";testValueName="overLimit";async=true
    PASS [expected FAIL] subtest: :limitTest="atMaximum";testValueName="overLimit";async=false
    PASS [expected FAIL] subtest: :limitTest="atMaximum";testValueName="overLimit";async=true
```

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `taplo format`.
- [x] Run `cargo clippy`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
  - [ ] `--target wasm32-unknown-emscripten`
- [x] Run `cargo xtask test` to run tests.
- [ ] Add change to `CHANGELOG.md`. See simple instructions inside file.

### CODE REVIEW

**Comment by sagudev** [comment]
@abey79 could you test if this raises error where you expected to?

**Comment by abey79** [comment]
TL;DR: yes, my offending code now panics 👍🏻 

**edit**: on second thought, that message seems wrong: the stride is 8 bytes and the attributes are 32 bytes in total, which is indeed non-compliant (the stride is allowed to be greater, not smaller).


Can be reproduced with this branch (https://github.com/abey79/vsvg/tree/antoine/wgpu-fix-test), which patches wgpu to this commit.

```
❯ cargo run -p whiskers-web-demo             
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.11s
     Running `target/debug/whiskers-web-demo`
thread 'main' panicked at /Users/hhip/.cargo/git/checkouts/wgpu-53e70f8674b08dd4/b92e624/wgpu/src/backend/wgpu_core.rs:1302:26:
wgpu error: Validation Error

Caused by:
  In Device::create_render_pipeline, label = 'line pipeline'
    Vertex buffer 0 stride 32 exceeds the limit 8


note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at library/core/src/panicking.rs:221:5:
panic in a function that cannot unwind
stack backtrace:
   0:        0x1015b9e8c - std::backtrace_rs::backtrace::libunwind::trace::hbebc8679d47bdc2c
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/../../backtrace/src/backtrace/libunwind.rs:116:5
   1:        0x1015b9e8c - std::backtrace_rs::backtrace::trace_unsynchronized::h3a2e9637943241aa
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:        0x1015b9e8c - std::sys::backtrace::_print_fmt::he430849680584674
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/sys/backtrace.rs:65:5
   3:        0x1015b9e8c - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h243268f17d714c7f
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/sys/backtrace.rs:40:26
   4:        0x1015da168 - core::fmt::rt::Argument::fmt::h0d339881c25f3c31
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/fmt/rt.rs:173:76
   5:        0x1015da168 - core::fmt::write::hb3cfb8a30e72d7ff
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/fmt/mod.rs:1182:21
   6:        0x1015b6e8c - std::io::Write::write_fmt::hfb2314975de9ecf1
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/io/mod.rs:1827:15
   7:        0x1015bafec - std::sys::backtrace::BacktraceLock::print::he14461129ccbfef5
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/sys/backtrace.rs:43:9
   8:        0x1015bafec - std::panicking::default_hook::{{closure}}::h14c7718ccf39d316
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/panicking.rs:269:22
   9:        0x1015bac10 - std::panicking::default_hook::hc62e60da3be2f352
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/panicking.rs:296:9
  10:        0x1015bbab0 - std::panicking::rust_panic_with_hook::h09e8a656f11e82b2
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/panicking.rs:800:13
  11:        0x1015bb3d8 - std::panicking::begin_panic_handler::{{closure}}::h1230eb3cc91b241c
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/panicking.rs:667:13
  12:        0x1015ba318 - std::sys::backtrace::__rust_end_short_backtrace::hc3491307aceda2c2
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/sys/backtrace.rs:168:18
  13:        0x1015bb0c8 - rust_begin_unwind
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/panicking.rs:665:5
  14:        0x10164dbbc - core::panicking::panic_nounwind_fmt::runtime::h5290ab2b4897aadc
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/panicking.rs:112:18
  15:        0x10164dbbc - core::panicking::panic_nounwind_fmt::h91ee161184879b56
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/panicking.rs:122:5
  16:        0x10164dc34 - core::panicking::panic_nounwind::heab7ebe7a6cd845c
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/panicking.rs:221:5
  17:        0x10164ddac - core::panicking::panic_cannot_unwind::hedc43d82620205bf
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/panicking.rs:309:5
  18:        0x1008ff94c - winit::platform_impl::macos::app_state::ApplicationDelegate::app_did_finish_launching::h26a705a9210618bb
                               at /Users/hhip/.cargo/registry/src/index.crates.io-6f17d22bba15001f/objc2-0.5.2/src/macros/declare_class.rs:981:25
  19:        0x18767d360 - <unknown>
  20:        0x18770e20c - <unknown>
  21:        0x18770e154 - <unknown>
  22:        0x18764bf9c - <unknown>
  23:        0x188805fd4 - <unknown>
  24:        0x18b1f9c78 - <unknown>
  25:        0x18b1f9a28 - <unknown>
  26:        0x18b1f7e88 - <unknown>
  27:        0x18b1f7a90 - <unknown>
  28:        0x18882ec8c - <unknown>
  29:        0x18882ea84 - <unknown>
thread caused non-unwinding panic. aborting.
zsh: abort      cargo run -p whiskers-web-demo
```

**Comment by abey79** [comment]
Repeating my edit above for visibility: on second thought, that message seems wrong. The stride is 8 bytes and the attributes are 32 bytes in total, which is indeed non-compliant (the stride is allowed to be greater, not smaller).

File: wgpu-core/src/device/resource.rs [hunk 1]
```diff
@@ -2926,6 +2916,46 @@ impl Device {
...
+                self.limits.max_vertex_buffer_array_stride as u64
+            } else {
+                vb_state.array_stride
+            };
+            if last_stride > max_stride {
+                return Err(pipeline::CreateRenderPipelineError::VertexStrideTooLarge {
+                    index: i as u32,
+                    given: last_stride as u32,
+                    limit: max_stride as u32,
+                });
```
Comment (sagudev) [inline wgpu-core/src/device/resource.rs]: here we should say that attribute stride is too large, not vertex stride

File: wgpu-core/src/device/resource.rs [hunk 2]
```diff
@@ -2926,6 +2916,48 @@ impl Device {
...
+            };
+            if last_stride > max_stride {
+                return Err(
+                    pipeline::CreateRenderPipelineError::VertexAttributeStrideTooLarge {
+                        index: i as u32,
+                        given: last_stride as u32,
+                        limit: max_stride as u32,
+                    },
+                );
+            }
```
Comment (sagudev) [inline wgpu-core/src/device/resource.rs]: We could do this check for each attribute (that would match the spec even more), which would give us better errors, with the small price of performance.

File: wgpu-core/src/device/resource.rs [hunk 3]
```diff
@@ -2926,6 +2916,48 @@ impl Device {
                     stride: vb_state.array_stride,
                 });
             }
+            let mut last_stride = 0;
+            for attribute in vb_state.attributes.iter() {
+                if attribute.offset % attribute.format.size().min(4) != 0 {
```
Comment (cwfitzgerald) [inline wgpu-core/src/device/resource.rs]:
```suggestion
                let required_offset_alignment = attribute.format.size().min(4);
                if (attribute.offset % required_offset_alignment) != 0 {
```

just a bit clearer, there's a lot going on here

File: wgpu-core/src/device/resource.rs (hunk 2 above)
Comment (cwfitzgerald) [reply wgpu-core/src/device/resource.rs]: This would be great - it would be a much better error and shouldn't be too expensive.

**Review by cwfitzgerald — CHANGES_REQUESTED** [review]
Looks good, some comments

File: wgpu-core/src/device/resource.rs (hunk 2 above)
Comment (sagudev) [reply wgpu-core/src/device/resource.rs]: Done in https://github.com/gfx-rs/wgpu/pull/6804/commits/3b56f03705493e849a7013414266c3fafa7a862d

**Review by cwfitzgerald — APPROVED** [review]
Nice!

**Comment by bschwind** [comment]
I'm running into this validation error with an instanced line renderer:

https://github.com/bschwind/simple-game/blob/3bd52ba7c64f1bd0c79066c6a73f61df15889745/src/graphics/lines2d.rs#L106-L113

The general technique is outlined here:

https://wwwtyro.net/2019/11/18/instanced-lines.html

Is it correct that I should migrate this code to:

* Modify the "point A" `wgpu::VertexBufferLayout` to just hold a single `Float32x3`
* Add another `wgpu::VertexBufferLayout` to hold "point B", another `Float32x3`
* When rendering, bind the _same_ instance buffer to shader location `1` and `2`, where the binding for `2` has an offset of one element in the buffer?

That seems to work, but I just wanted to double check.

**Comment by cwfitzgerald** [comment]
Yeah that sounds good to me!

**Comment by bschwind** [comment]
@cwfitzgerald thanks! Glad there's an easy workaround that doesn't require maintaining an extra buffer or something like that.
